strings/longest_palindromic_substring.py

```diff
@@ -0,0 +1,55 @@
+"""
+1. Maintain a variable ‘ longest_length = 1 ‘
+(for storing longest palindromic substring length) and ‘ start =0 ‘.
+2.	The idea is very simple, we will traverse through the entire
+string with i=0 to i<(length of string) while traversing, initialize
+’start‘ and ‘end‘ pointer such that start= i-1 and end= i+1.
+3.	keep incrementing ‘end’ until str[end]==str[i]  similarly
+keep decrementing ‘start’ until str[start]==str[i].
+4.	Finally we will keep incrementing ‘end’ and decrementing
+‘start’ until str[start]==str[end].
+5.	calculate length=end-start-1, if length > longest_length
+then longest_length = length and start = start+1 .
+6.	Return the longest palindromic substring.
+"""
+
+
+def longest_palindromic_substring(string: str) -> str:
+    """
+    This program takes O(n^2) time complexity to find the longest palindromic substring
+    This program takes O(1) space complexity to find the longest palindromic substring
```

Comment:
How do we prove these two statements?

Reply:
Apart from variables this program is not using any extra space so O(1) space complexity.  Coming to time complexity, we will are traversing the entire string with i=0 to i<(length of string) and for every character 
1. We first traverse the left side of character by decrementing 'start'.
2. Then we traverse the right side of character by incrementing 'end
3.  finally we will keep incrementing ‘end’ and decrementing ‘start’ until str[start]==str[end]

Lets take an example,
string = bcccb, if we are index 2
1. we first traverse left side until string[i]==string[start], so start becomes 0
2. we traverse right side until string[i]==string[end], end becomes 4
3. finally we will keep incrementing ‘end’ and decrementing ‘start’ until str[start]==str[end], start becomes -1, end becomes 5

In worst case if the given string contains all similar characters, Eg: 'aaaaaa' then for every character we have to traverse entire string. In this case time complexity becomes O(n^2)

Reply:
> https://github.com/TheAlgorithms/Python/blob/master/strings/is_palindrome.py#L20

this program is for finding whether the given string is palindrome or not it can be done in O(n) by just checking with the reverse of the string. But In our case it is longest palindromic substring.